From 6b55c9ed8494c44a012f013e530fc20cac4ec2b0 Mon Sep 17 00:00:00 2001
From: Pawan Gupta <pawan.kumarx.gupta@intel.com>
Date: Fri, 10 Jun 2016 16:30:51 -0700
Subject: [PATCH 416/441] tty: Merge Puma6 and Puma7

Merged puma6 specific changes with puma7
Puma6 changes goes under CONFIG_GEN3_UART

Change-Id: I1e753b272e9c6afd594732cb85c485a4587d51df
Signed-off-by: Pawan Gupta <pawan.kumarx.gupta@intel.com>
---
 drivers/tty/serial/8250/8250_core.c | 49 ++++++++++++++++++++++++++++++++++++-
 1 file changed, 48 insertions(+), 1 deletion(-)

--- a/drivers/tty/serial/8250/8250_core.c
+++ b/drivers/tty/serial/8250/8250_core.c
@@ -45,6 +45,10 @@
 #include <asm/io.h>
 #include <asm/irq.h>
 
+#ifdef CONFIG_GEN3_UART
+#include <linux/pci.h>
+#endif
+
 #include "8250.h"
 
 /*
@@ -58,6 +62,13 @@ static unsigned int nr_uarts = CONFIG_SE
 
 static struct uart_driver serial8250_reg;
 
+/*
+ * Intel CE2600 have only two UARTS, while previous platform have three UARTS
+*/
+#ifdef CONFIG_GEN3_UART
+#define CE2600_SERIAL_8250_NR_UARTS 2
+#endif
+
 static int serial_index(struct uart_port *port)
 {
 	return (serial8250_reg.minor - 64) + port->line;
@@ -233,10 +244,20 @@ static const struct serial8250_config ua
 		.flags		= UART_CAP_FIFO | UART_NATSEMI,
 	},
 	[PORT_XSCALE] = {
+/*
+ * The following code is for Intel Media SOC Gen3 base support.
+*/
+#ifdef CONFIG_GEN3_UART
+		.name		= "GEN3_serial",
+		.fifo_size	= 64,
+		.tx_loadsz	= 64,
+		.fcr		= UART_FCR_ENABLE_FIFO | UART_FCR_DMA_SELECT | UART_FCR_R_TRIG_10,
+#else
 		.name		= "XScale",
 		.fifo_size	= 32,
 		.tx_loadsz	= 32,
 		.fcr		= UART_FCR_ENABLE_FIFO | UART_FCR_R_TRIG_10,
+#endif
 		.flags		= UART_CAP_FIFO | UART_CAP_UUE | UART_CAP_RTOIE,
 	},
 	[PORT_OCTEON] = {
@@ -1045,6 +1066,9 @@ static void autoconfig(struct uart_8250_
 	struct uart_port *port = &up->port;
 	unsigned long flags;
 	unsigned int old_capabilities;
+#ifdef CONFIG_GEN3_UART
+	unsigned int id;
+#endif
 
 	if (!port->iobase && !port->mapbase && !port->membase)
 		return;
@@ -1059,6 +1083,13 @@ static void autoconfig(struct uart_8250_
 	spin_lock_irqsave(&port->lock, flags);
 
 	up->capabilities = 0;
+#ifdef CONFIG_GEN3_UART
+	/* do not enable modem status interrupt for IntelCE uart0 port */
+	intelce_get_soc_info(&id, NULL);
+	if((CE4200_SOC_DEVICE_ID == id) && (0 == serial_index(&up->port)))
+		up->bugs = UART_BUG_NOMSR;
+	else
+#endif
 	up->bugs = 0;
 
 	if (!(port->flags & UPF_BUGGY_UART)) {
@@ -1908,11 +1939,21 @@ static void serial8250_put_poll_char(str
 	 *	First save the IER then disable the interrupts
 	 */
 	ier = serial_port_in(port, UART_IER);
+
+	/*
+	 * The following code is for Intel Media SOC Gen3 base support.
+	*/
+#ifdef CONFIG_GEN3_UART
+	/*
+	 * Should enable UUE (Uart Unit Enable) bit.
+	*/
+		serial_port_out(port, UART_IER, UART_IER_UUE);
+#else
 	if (up->capabilities & UART_CAP_UUE)
 		serial_port_out(port, UART_IER, UART_IER_UUE);
 	else
 		serial_port_out(port, UART_IER, 0);
-
+#endif
 	wait_for_xmitr(up, BOTH_EMPTY);
 	/*
 	 *	Send the character out.
@@ -3321,6 +3362,12 @@ static int __init serial8250_init(void)
 {
 	int ret;
 
+#ifdef CONFIG_GEN3_UART
+	unsigned int id;
+	intelce_get_soc_info(&id, NULL);
+	if(CE2600_SOC_DEVICE_ID == id)
+		nr_uarts = CE2600_SERIAL_8250_NR_UARTS;
+#endif
 	serial8250_isa_init_ports();
 
 	printk(KERN_INFO "Serial: 8250/16550 driver, "
